# Python Foundations — Part A: The Basics

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/tech4alltraining/aiml/blob/main/mlai-genai-internship/notebooks/00a_python_foundations.ipynb)

**ML/AI & GenAI Internship** · For students who have never written Python. Each concept: examples → 5 exercises → quiz → tasks.

---

## Read this first

**This notebook assumes you have never written a line of Python.** That is fine. Nothing here needs maths beyond arithmetic.

Seven concepts, each in the same shape:

```text
📘 Examples          2-3 worked examples you run and read
✏️ Practice now      5 short exercises, with solutions
🧠 Quick quiz        5 questions to check you understood
🎯 Tasks             2-3 longer pieces for after the session
```

| # | Concept |
|---|---|
| 1 | Printing and variables |
| 2 | Data types and conversion |
| 3 | Operators |
| 4 | Strings |
| 5 | Collections: list, tuple, set, dict |
| 6 | Conditions and loops |
| 7 | Functions |

**How to run a cell:** click it and press `Shift+Enter`.

**When you hit an error** — and you will — read the **last line first**. That line names the actual problem. Everything above it is just the path the computer took to get there. Errors are not failure; they are the computer telling you precisely what it needs.

---

# Concept 1 — Printing and variables

🧠 **Analogy: labelled boxes.**

A variable is a labelled box you put a value in. Write `age = 21` and you have a box labelled `age` containing 21. Later you just say `age` and Python fetches what is inside.

Two rules that trip up beginners:

- **The `=` sign is not "equals".** It means *put this value in this box*. `x = 5` is an instruction, not a statement of fact.
- **The box keeps the last thing you put in it.** `x = 5` then `x = 9` leaves 9. The 5 is gone.

## 📘 Examples

### Example 1 — printing

In [1]:
# print() shows something on the screen.
# Anything in quotes is TEXT and is printed exactly as written.

print("Hello, World!")
print("Python is fun!")

# Numbers do not need quotes.
print(42)
print(3.14)

# print() with several things separated by commas puts a space between them.
print("My age is", 21)

Hello, World!
Python is fun!
42
3.14
My age is 21


### Example 2 — variables

In [2]:
# Create a variable: give the box a name, then put a value in it.
name = "Asha"
age = 21
height = 5.4
is_student = True

# Fetch what is inside by using the name.
print(name)
print(age)

# A variable keeps only the LAST value you put in it.
score = 50
score = 87
print("score is", score)     # 87 - the 50 is gone

# You can use a variable to make a new one.
next_year = age + 1
print("next year:", next_year)

Asha
21
score is 87
next year: 22


### Example 3 — f-strings, the readable way to print

In [3]:
name = "Asha"
marks = 87.456

# The clumsy way, with commas:
print("Student", name, "scored", marks)

# The f-string way: put an f before the quote, then {variable} inside.
print(f"Student {name} scored {marks}")

# f-strings can format numbers too. :.1f means "1 decimal place".
print(f"Student {name} scored {marks:.1f}")
print(f"Student {name} scored {marks:.0f}")

# You can do arithmetic inside the braces.
print(f"{name} needs {100 - marks:.1f} more marks for full marks")

Student Asha scored 87.456
Student Asha scored 87.456
Student Asha scored 87.5
Student Asha scored 87
Asha needs 12.5 more marks for full marks


## ✏️ Practice now — Printing and variables

Five short exercises. Do them **before** moving on — this is where the learning happens.

**1.** Print the text `Learning Python` on one line.

**2.** Create a variable `college` holding your college name, and print it.

**3.** Create `subject = "Machine Learning"` and `hours = 40`. Print `I am studying Machine Learning for 40 hours` using an f-string.

**4.** Create `price = 249.789`. Print it rounded to **2 decimal places** using an f-string.

**5.** Create `a = 10`, then change `a` to `25`, then print `a`. What is printed, and why?

In [4]:
# 1.


# 2.


# 3.


# 4.


# 5.

<details>
<summary><b>Solutions — open only after you have tried all five</b></summary>

```python
# 1
print("Learning Python")

# 2
college = "NIT Calicut"
print(college)

# 3
subject = "Machine Learning"
hours = 40
print(f"I am studying {subject} for {hours} hours")

# 4
price = 249.789
print(f"{price:.2f}")

# 5
a = 10
a = 25
print(a)      # 25 - a variable keeps only the LAST value assigned to it
```

</details>

## 🧠 Quick quiz — Printing and variables

Answer all five from memory. No scrolling back.

**Q1.** What does `=` mean in Python?

**Q2.** `x = 5` then `x = 9`. What does `print(x)` show?

**Q3.** What is the difference between `print(age)` and `print("age")`?

**Q4.** What does the `f` do in `print(f"Hello {name}")`?

**Q5.** What does `{price:.2f}` do inside an f-string?

<details>
<summary><b>Answers</b></summary>

**A1.** **Put this value in this box.** It is an instruction, not a statement of equality. `x = 5` tells Python to store 5 in a box named `x`.

**A2.** `9`. A variable holds only the **last** value assigned to it. The 5 is gone.

**A3.** `print(age)` shows the **value inside** the variable, for example 21. `print("age")` shows the literal text `age`, because anything in quotes is treated as text.

**A4.** It makes it an **f-string**, so anything inside `{ }` is replaced by the value of that expression. Without the `f`, Python prints the braces literally: `Hello {name}`.

**A5.** Formats the number to **2 decimal places**. `:.0f` gives no decimals, `:.3f` gives three. Handy for money and for keeping output tidy.

</details>

## 🎯 Tasks — Printing and variables

Longer work. Do these after the session, in your own time.

### Task 1: Your ID card

Create variables for: your name, age, course, semester, and CGPA.

Print a tidy ID card using f-strings, like this:

```text
==============================
 NAME     : Asha Menon
 AGE      : 21
 COURSE   : Computer Science
 SEMESTER : 6
 CGPA     : 8.45
==============================
```

Hint: `:<10` left-aligns text in a 10-character space. Try `f"{'NAME':<10}"`.

### Task 2: Hospital patient record

A hospital checks in a patient: name `John Doe`, age 30, blood group `O+`, weight 72.5 kg, and they are a new patient.

1. Create a variable for each piece of information
2. Print each one with its type (use `type(x)`)
3. Print a one-line summary using an f-string
4. The patient's weight is re-measured as 71.8. Update the variable and print the summary again

---

# Concept 2 — Data types and conversion

🧠 **Analogy: what kind of thing is in the box?**

A box can hold text, a whole number, a decimal, or a yes/no. Python cares which, because **you can add two numbers but you cannot sensibly add two yes/nos.**

| Type | Called | Example | What it is |
|---|---|---|---|
| `str` | string | `"Asha"` | Text. **Always in quotes** |
| `int` | integer | `21` | A whole number |
| `float` | float | `5.4` | A decimal number |
| `bool` | boolean | `True` | Yes or no. Capital T, capital F |

> ⚠️ **`"21"` and `21` are different things.** The first is text that happens to look like a number. `"21" + "5"` gives `"215"` — it glues the text together. `21 + 5` gives `26`.

## 📘 Examples

### Example 1 — checking the type

In [5]:
name = "Asha"
age = 21
height = 5.4
is_student = True

# type() tells you what kind of thing a variable holds.
print(type(name))
print(type(age))
print(type(height))
print(type(is_student))

# .__name__ prints it more tidily
print()
for value in [name, age, height, is_student]:
    print(f"{str(value):<10} is a {type(value).__name__}")

<class 'str'>
<class 'int'>
<class 'float'>
<class 'bool'>

Asha       is a str
21         is a int
5.4        is a float
True       is a bool


### Example 2 — the text-vs-number trap

In [6]:
text_number = "21"      # text that LOOKS like a number
real_number = 21        # an actual number

print("21 + 5 as text  :", text_number + "5")     # glues them: "215"
print("21 + 5 as numbers:", real_number + 5)      # adds them: 26

# This is the single most common beginner error:
try:
    print(text_number + 5)
except TypeError as error:
    print("\nERROR:", error)
    print("Python will not guess. You must convert first.")

21 + 5 as text  : 215
21 + 5 as numbers: 26

ERROR: can only concatenate str (not "int") to str
Python will not guess. You must convert first.


### Example 3 — converting between types

In [7]:
# int()   -> whole number
# float() -> decimal
# str()   -> text
# bool()  -> True/False

print(int("21") + 5)          # 26  - text to number, then add
print(float("3.14") * 2)      # 6.28
print("Age: " + str(21))      # "Age: 21" - number to text, then glue

# int() on a float CHOPS OFF the decimal, it does not round.
print(int(7.9))               # 7, not 8
print(round(7.9))             # 8  - use round() if you want rounding

# Converting nonsense fails, and the error tells you exactly why.
try:
    int("hello")
except ValueError as error:
    print("\nERROR:", error)

26
6.28
Age: 21
7
8

ERROR: invalid literal for int() with base 10: 'hello'


## ✏️ Practice now — Data types

Five short exercises. Do them **before** moving on — this is where the learning happens.

**1.** Create one variable of each type (`str`, `int`, `float`, `bool`) and print each with its type.

**2.** What is `"7" + "3"`? What is `7 + 3`? Print both and explain the difference in a comment.

**3.** Convert the text `"45"` to a number and add 5 to it.

**4.** Convert `9.99` to an integer. What happened to the `.99`?

**5.** Try `int("12.5")`. Read the error. Then fix it so it works.

In [8]:
# 1.


# 2.


# 3.


# 4.


# 5.

<details>
<summary><b>Solutions — open only after you have tried all five</b></summary>

```python
# 1
s, i, f, b = "Asha", 21, 5.4, True
for v in [s, i, f, b]:
    print(f"{str(v):<8} {type(v).__name__}")

# 2
print("7" + "3")   # "73"  - text is GLUED together
print(7 + 3)       # 10    - numbers are ADDED

# 3
print(int("45") + 5)      # 50

# 4
print(int(9.99))          # 9 - int() CHOPS the decimal off, it does not round
print(round(9.99))        # 10

# 5
# int("12.5") fails: "12.5" is not a whole number in text form.
print(int(float("12.5")))   # convert to float FIRST, then to int -> 12
```

</details>

## 🧠 Quick quiz — Data types

Answer all five from memory. No scrolling back.

**Q1.** What is the difference between `"21"` and `21`?

**Q2.** `int(7.9)` gives 7, not 8. Why?

**Q3.** Why does `int("12.5")` fail when `int("12")` works?

**Q4.** You read a number from a user and adding 5 gives an error. What is wrong?

**Q5.** What does `type(True)` return, and what is unusual about writing `True` in Python?

<details>
<summary><b>Answers</b></summary>

**A1.** `"21"` is a **string** — text that happens to look like a number. `21` is an **integer**. `"21" + "5"` glues them into `"215"`; `21 + 5` adds them to 26.

**A2.** `int()` **chops off** the decimal part rather than rounding. If you want 8, use `round(7.9)`. This catches people out when converting averages.

**A3.** `int()` on a string expects a whole number in text form. `"12.5"` has a decimal point, so it is not a valid integer string. Convert in two steps: `int(float("12.5"))`.

**A4.** `input()` always returns a **string**, even when the user typed digits. Convert it first: `age = int(input(...))`. This is the single most common beginner error in Python.

**A5.** `bool`. The unusual part is the **capital T** — `true` in lowercase is not valid Python and gives a `NameError`. Same for `False`.

</details>

## 🎯 Tasks — Data types

Longer work. Do these after the session, in your own time.

### Task 1: The type detective

Write a program that takes a list of mixed values and reports what each one is:

```python
values = [42, "42", 42.0, True, "hello", 0, "", None, [1, 2]]
```

For each value print: the value, its type, and whether `bool(value)` is True or False.

**Then answer:** which values are 'falsy' (become `False`)? There are four. Why do you think those four were chosen?

### Task 2: Safe converter

Write a program that tries to convert a list of strings to numbers, handling the ones that cannot be converted.

```python
inputs = ["45", "3.14", "hello", "-7", "12.5", "", "1e3"]
```

For each: print either the converted number and its type, or a friendly message saying why it failed.

Use `try` / `except ValueError`. **Do not let the program crash.**

---

# Concept 3 — Operators

Operators are the symbols that *do* things: add, compare, combine.

| Group | Operators | What they give back |
|---|---|---|
| **Arithmetic** | `+  -  *  /  //  %  **` | A number |
| **Comparison** | `==  !=  >  <  >=  <=` | `True` or `False` |
| **Logical** | `and  or  not` | `True` or `False` |
| **Membership** | `in`, `not in` | `True` or `False` |

> ⚠️ **`=` and `==` are completely different.** `=` puts a value in a box. `==` asks "are these the same?" Mixing them up is the most common typo in programming.

## 📘 Examples

### Example 1 — arithmetic, including the three people forget

In [9]:
a, b = 17, 5

print(f"{a} + {b}  = {a + b}")
print(f"{a} - {b}  = {a - b}")
print(f"{a} * {b}  = {a * b}")
print(f"{a} / {b}  = {a / b}      <- normal division ALWAYS gives a float")
print(f"{a} // {b} = {a // b}          <- floor division: whole part only")
print(f"{a} % {b}  = {a % b}          <- modulus: the REMAINDER")
print(f"{a} ** 2 = {a ** 2}         <- power")

print("\nWhy % matters: it tells you if a number divides evenly.")
print(f"  10 % 2 = {10 % 2}  -> 0 means 10 is EVEN")
print(f"  11 % 2 = {11 % 2}  -> 1 means 11 is ODD")

17 + 5  = 22
17 - 5  = 12
17 * 5  = 85
17 / 5  = 3.4      <- normal division ALWAYS gives a float
17 // 5 = 3          <- floor division: whole part only
17 % 5  = 2          <- modulus: the REMAINDER
17 ** 2 = 289         <- power

Why % matters: it tells you if a number divides evenly.
  10 % 2 = 0  -> 0 means 10 is EVEN
  11 % 2 = 1  -> 1 means 11 is ODD


### Example 2 — comparison gives True or False

In [10]:
marks = 78

print(f"marks == 78 -> {marks == 78}    are they equal?")
print(f"marks != 78 -> {marks != 78}   are they different?")
print(f"marks >= 75 -> {marks >= 75}    at least 75?")
print(f"marks <  50 -> {marks < 50}   less than 50?")

# The classic mistake:
print("\n=  puts a value in a box")
print("== asks whether two things are the same")

x = 5          # assignment
print(x == 5)  # comparison -> True

marks == 78 -> True    are they equal?
marks != 78 -> False   are they different?
marks >= 75 -> True    at least 75?
marks <  50 -> False   less than 50?

=  puts a value in a box
== asks whether two things are the same
True


### Example 3 — logical and membership operators

In [11]:
marks = 78
attendance = 65

# and -> BOTH must be true
print("passed and good attendance:", marks >= 40 and attendance >= 75)

# or -> AT LEAST ONE must be true
print("passed or good attendance :", marks >= 40 or attendance >= 75)

# not -> flips it
print("did NOT pass              :", not (marks >= 40))

# in -> is this inside that?
subjects = ["Maths", "Physics", "Python"]
print("\n'Python' in subjects   :", "Python" in subjects)
print("'History' in subjects  :", "History" in subjects)
print("'y' in 'Python'        :", "y" in "Python")     # works on text too

passed and good attendance: False
passed or good attendance : True
did NOT pass              : False

'Python' in subjects   : True
'History' in subjects  : False
'y' in 'Python'        : True


## ✏️ Practice now — Operators

Five short exercises. Do them **before** moving on — this is where the learning happens.

**1.** Print the result of `23 // 4` and `23 % 4`. Explain in a comment what each means.

**2.** Write an expression that is `True` when a number `n` is even.

**3.** Given `age = 20` and `has_id = True`, write one expression that is `True` only when **both** are satisfied (age at least 18 **and** has ID).

**4.** Given `fruits = ["apple", "banana"]`, check whether `"mango"` is in it.

**5.** Predict the output of `print(5 == 5.0)` before running it. Were you right?

In [12]:
# 1.


# 2.


# 3.


# 4.


# 5.

<details>
<summary><b>Solutions — open only after you have tried all five</b></summary>

```python
# 1
print(23 // 4)   # 5 - how many whole 4s fit into 23
print(23 % 4)    # 3 - what is left over

# 2
n = 10
print(n % 2 == 0)      # True when n is even

# 3
age, has_id = 20, True
print(age >= 18 and has_id)

# 4
fruits = ["apple", "banana"]
print("mango" in fruits)     # False

# 5
print(5 == 5.0)   # True - Python compares the VALUE, not the type
```

</details>

## 🧠 Quick quiz — Operators

Answer all five from memory. No scrolling back.

**Q1.** What is the difference between `/` and `//`?

**Q2.** What does `%` do, and what is it most often used for?

**Q3.** What is the difference between `=` and `==`?

**Q4.** `marks >= 40 and attendance >= 75`. When is this True?

**Q5.** What does `"y" in "Python"` return, and what does that tell you about `in`?

<details>
<summary><b>Answers</b></summary>

**A1.** `/` is normal division and **always** returns a float: `10 / 2` gives `5.0`. `//` is floor division and keeps only the whole part: `17 // 5` gives `3`.

**A2.** It gives the **remainder** after division. Most commonly used to test divisibility: `n % 2 == 0` is True when `n` is even.

**A3.** `=` **assigns** — puts a value in a variable. `==` **compares** — asks whether two things are equal and returns True or False. Confusing them is the most common typo in programming.

**A4.** Only when **both** sides are True. `and` requires everything; `or` requires at least one.

**A5.** `True`. It tells you `in` works on **text** as well as lists — it checks whether one string appears inside another.

</details>

## 🎯 Tasks — Operators

Longer work. Do these after the session, in your own time.

### Task 1: Simple calculator

Write a calculator that:

1. Stores two numbers in variables
2. Prints all seven arithmetic results (`+ - * / // % **`), each labelled
3. Prints whether the first is larger, smaller, or equal to the second
4. Prints whether both are even, using `and`
5. Handles division by zero without crashing

Test it with `(17, 5)`, `(10, 0)` and `(-8, 3)`. **Does negative floor division do what you expected?** Look up why.

### Task 2: Eligibility checker

A scholarship requires: CGPA at least 8.0, attendance at least 75%, no backlogs, and family income under 500000.

1. Create a variable for each condition
2. Write **one** expression combining all four that gives `True` or `False`
3. Print a clear message for the student
4. Then print **which specific conditions failed**, so the message is actually useful

Point 4 is the difference between a program and a useful program.

---

# Concept 4 — Strings

A **string** is text. In ML you meet strings constantly: column names, categories, user reviews, prompts.

🧠 **Analogy: a row of numbered letter tiles.** Each character has a position, starting at **0**. You can look at one tile, take a run of tiles, or count them.

## 📘 Examples

### Example 1 — indexing and slicing

In [13]:
text = "Machine Learning"

#  M  a  c  h  i  n  e     L  e  a  r  n  i  n  g
#  0  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15

print("text        :", text)
print("length      :", len(text))
print("text[0]     :", text[0], "   first character - counting starts at 0")
print("text[-1]    :", text[-1], "   last character - negative counts backwards")
print("text[0:7]   :", text[0:7], " a SLICE: includes 0, EXCLUDES 7")
print("text[8:]    :", text[8:], "  from 8 to the end")
print("text[:7]    :", text[:7], " from the start to 6")

text        : Machine Learning
length      : 16
text[0]     : M    first character - counting starts at 0
text[-1]    : g    last character - negative counts backwards
text[0:7]   : Machine  a SLICE: includes 0, EXCLUDES 7
text[8:]    : Learning   from 8 to the end
text[:7]    : Machine  from the start to 6


### Example 2 — the string methods you will actually use

In [14]:
messy = "  Machine Learning IS Fun!  "

print(f"original      : {messy!r}")
print(f".strip()      : {messy.strip()!r}        <- removes surrounding spaces")
print(f".lower()      : {messy.lower().strip()!r}")
print(f".upper()      : {messy.upper().strip()!r}")
print(f".replace()    : {messy.strip().replace('Fun', 'Useful')!r}")
print(f".split()      : {messy.split()}")
print(f".startswith() : {messy.strip().startswith('Machine')}")
print(f"count of 'n'  : {messy.lower().count('n')}")

# .strip() and .lower() are how you clean messy category data:
raw = ["  Delhi", "delhi ", "DELHI", "Delhi"]
cleaned = [city.strip().lower() for city in raw]
print(f"\nraw     : {raw}")
print(f"cleaned : {cleaned}   <- now they all match")

original      : '  Machine Learning IS Fun!  '
.strip()      : 'Machine Learning IS Fun!'        <- removes surrounding spaces
.lower()      : 'machine learning is fun!'
.upper()      : 'MACHINE LEARNING IS FUN!'
.replace()    : 'Machine Learning IS Useful!'
.split()      : ['Machine', 'Learning', 'IS', 'Fun!']
.startswith() : True
count of 'n'  : 4

raw     : ['  Delhi', 'delhi ', 'DELHI', 'Delhi']
cleaned : ['delhi', 'delhi', 'delhi', 'delhi']   <- now they all match


### Example 3 — joining and splitting

In [15]:
sentence = "Python is a great language for machine learning"

words = sentence.split()          # split on spaces -> a list
print("words:", words)
print("how many:", len(words))

# join() does the opposite: a list back into one string.
print("\njoined with '-':", "-".join(words))
print("joined with ', ':", ", ".join(words[:3]))

# Splitting a CSV line - you will do this constantly.
row = "Asha,21,Computer Science,8.45"
fields = row.split(",")
print(f"\nfields: {fields}")
print(f"name is {fields[0]}, CGPA is {fields[3]}")

words: ['Python', 'is', 'a', 'great', 'language', 'for', 'machine', 'learning']
how many: 8

joined with '-': Python-is-a-great-language-for-machine-learning
joined with ', ': Python, is, a

fields: ['Asha', '21', 'Computer Science', '8.45']
name is Asha, CGPA is 8.45


## ✏️ Practice now — Strings

Five short exercises. Do them **before** moving on — this is where the learning happens.

**1.** Given `s = "Data Science"`, print its length, first character and last character.

**2.** Print `"Data"` from that string using a slice.

**3.** Convert `"  HELLO world  "` to lowercase with no surrounding spaces.

**4.** Count how many words are in `"machine learning is fun and useful"`.

**5.** Given `row = "Ravi,19,Physics"`, split it and print `Ravi is 19 and studies Physics` using an f-string.

In [16]:
# 1.


# 2.


# 3.


# 4.


# 5.

<details>
<summary><b>Solutions — open only after you have tried all five</b></summary>

```python
# 1
s = "Data Science"
print(len(s), s[0], s[-1])

# 2
print(s[0:4])       # or s[:4]

# 3
print("  HELLO world  ".strip().lower())

# 4
print(len("machine learning is fun and useful".split()))

# 5
row = "Ravi,19,Physics"
name, age, subject = row.split(",")
print(f"{name} is {age} and studies {subject}")
```

</details>

## 🧠 Quick quiz — Strings

Answer all five from memory. No scrolling back.

**Q1.** In `text[0:7]`, is character 7 included?

**Q2.** What does `text[-1]` give you, and why is it useful?

**Q3.** Why would you call `.strip().lower()` on category data?

**Q4.** What does `"a,b,c".split(",")` return?

**Q5.** `"Hello".replace("l", "L")` — does the original string change?

<details>
<summary><b>Answers</b></summary>

**A1.** **No.** A slice includes the start and **excludes** the stop. `text[0:7]` gives characters 0 to 6 — seven characters in total.

**A2.** The **last** character. Negative indexes count backwards from the end, so you get the last item without needing to know the length.

**A3.** To make inconsistent values match. `"  Delhi"`, `"delhi "` and `"DELHI"` are three different strings to Python but the same city to a human. Cleaning them prevents your model treating one city as three.

**A4.** A **list**: `['a', 'b', 'c']`. `split` cuts a string into pieces; `join` does the reverse, gluing a list back into one string.

**A5.** **No.** Strings in Python are immutable: methods return a **new** string. You must capture it: `text = text.replace(...)`. Forgetting this is a very common bug.

</details>

## 🎯 Tasks — Strings

Longer work. Do these after the session, in your own time.

### Task 1: Text analyser

Given a paragraph of text, print:

1. Total characters, and characters excluding spaces
2. Total words
3. Number of **unique** words (case-insensitive)
4. The five longest words
5. How many times each vowel appears
6. The most common word

Test it on a paragraph from this notebook. **Then test it on a paragraph with punctuation** — does your word count still work? Fix it if not.

### Task 2: Clean the messy column

Real category data arrives like this:

```python
cities = ["  Delhi", "delhi ", "DELHI", "Mumbai", " mumbai", "Chennai",
          "chennai  ", "MUMBAI", "New  Delhi", "new delhi"]
```

1. How many *distinct* values does Python see right now?
2. Clean them so that identical cities match
3. How many distinct values now?
4. `"New  Delhi"` has a double space. Handle that too
5. Print a count of each cleaned city

**Why this matters:** a model treating `Delhi`, `delhi` and `DELHI` as three separate cities learns three weaker patterns instead of one strong one.

### Task 3: Build a name formatter

Write a program that takes messy names and formats them properly.

```python
names = ["  asha MENON ", "RAVI kumar", "meera  s  nair", "j. doe"]
```

Output should be `Asha Menon`, `Ravi Kumar`, `Meera S Nair`, `J. Doe`.

Handle: surrounding spaces, wrong capitalisation, double spaces between words. Look up `.title()` and think about where it goes wrong.